go-controller/pkg/clustermanager/networkconnect/cluster_network_connect.go

```diff
@@ -0,0 +1,280 @@
+package networkconnect
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	utilnet "k8s.io/utils/net"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	networkconnectv1 "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/clusternetworkconnect/v1"
+	apitypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/types"
+	ovntypes "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+)
+
+var (
+	errConfig = errors.New("configuration error")
+)
+
+func (c *Controller) reconcileClusterNetworkConnect(key string) error {
+	c.Lock()
+	defer c.Unlock()
+
+	startTime := time.Now()
+	_, cncName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	klog.V(5).Infof("reconcileClusterNetworkConnect %s", cncName)
+	defer func() {
+		klog.Infof("reconcileClusterNetworkConnect %s took %v", cncName, time.Since(startTime))
+	}()
+	cnc, err := c.cncLister.Get(cncName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	if cnc == nil {
+		// CNC is being deleted, clean up resources
+		cncState, exists := c.cncCache[cncName]
+		if exists {
+			// Release tunnel key
+			c.tunnelKeysAllocator.ReleaseKeys(cncName)
+			klog.V(4).Infof("Released tunnel key for deleted CNC %s", cncName)
+
+			// Release allocated subnets from this CNC's allocator
+			if cncState.allocator != nil {
+				cncState.allocator.ReleaseAllAllocations()
+				klog.V(4).Infof("Released all subnets for deleted CNC %s", cncName)
+			}
+		}
+
+		// Clean up the cache
+		delete(c.cncCache, cncName)
+		klog.V(4).Infof("Cleaned up cache for deleted CNC %s", cncName)
+		return nil
+	}
+
+	// STEP1: Validate the CNC
+	// STEP2: Discover the selected UDNs and CUDNs
+	discoveredLayer3Networks, discoveredLayer2Networks, err := c.discoverSelectedNetworks(cnc)
+	if err != nil {
+		return err
+	}
+	if len(discoveredLayer3Networks) == 0 && len(discoveredLayer2Networks) == 0 {
```

Comment:
we may need some cleanup in this case? if the network got un-selected?

Reply:
yup! this was one of the bits that I hadn't gotten around to finishing, but is now ready for review in my next push. PTAL.